WooExpress: Trial flow checks that WooCommerce is installed before completing

markbiek opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #73307

## Proposed Changes

Previously, we redirected to `/home` as soon as the Atomic transfer was complete. This lead to issues because, sometimes, WooCommerce wasn't finished installed before the redirect so we'd end up on the main dashboard instead of the WooCommerce launch checklist.

This adds a new step where we wait for WooCommerce to finish installing before we redirect.

## Testing Instructions

Prior to applying this patch, visit http://calypso.localhost:3000/setup/wooexpress/. In some cases, you may end up on `/home` instead of `/wp-admin/admin.php?page=wc-admin`. It seems slightly more likely to happen if you're sandboxed but you may have try multiple times to reproduce the behavior.

After applying the patch, you should always end up at `/wp-admin/admin.php?page=wc-admin`.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

markbiek committed:
WooExpress trial checks that WooCommerce is installed before completing

Previously, we redirected to `/home` as soon as the Atomic transfer was
complete. This lead to issues because, sometimes, WooCommerce wasn't
finished installed before the redirect so we'd end up on the main
dashboard instead of the WooCommerce launch checklist.

This adds a new step where we wait for WooCommerce to finish installing
before we redirect.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-plugin-install/index.tsx:
```diff
@@ -0,0 +1,120 @@
+			let stopPollingPlugins = false;
+
+			while ( ! stopPollingPlugins ) {
+				await wait( 3000 );
```
I think we should make this interval shorter so we don't introduce unnecessary delays of up to 3 seconds. Maybe 0.5 to 1 second intervals could work?

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-plugin-install/index.tsx:
```diff
@@ -0,0 +1,120 @@
+						if ( woocommercePlugin ) {
+							stopPollingPlugins = true;
+						}
+					}
```
This feels icky to me. Can we push the set of desired plugin slugs into the `ONBOARD_STORE` via a dedicated storage location/key? We could persist the set of plugins at the start of a flow (or during any step before this one), and simply have this step verify that a list of plugin slugs have been installed on the site.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-plugin-install/index.tsx:
```diff
@@ -0,0 +1,120 @@
+
+		submit?.();
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
This is a red flag to me: why are we disabling the check? At the very least, please add a comment to note why.

markbiek replied:
Most steps are using this to one degree or another, but I'll add a comment for clarity.

<img width="698" alt="image" src="https://user-images.githubusercontent.com/917632/219366043-d36e8928-fe49-4f72-be74-d50f44738f91.png">

daledupreez reviewed (commented):
This isn't working for me. All the API calls to `/sites/:siteSlug/lpugins` are returning a 403 status with the following error content:
```
{"error":"unauthorized","message":"This user is not authorized to activate_plugins on this blog."}
```
This is a weird catch-22, as we aren't allowing _anyone_ plugin activate or install permissions on the site. :-/

markbiek committed:
Shorten plugin polling interval

markbiek committed:
Add comment for react-hooks/exhaustive-deps disable

markbiek committed:
Use the ONBOARD_STORE for the list of plugins to verify

In the `siteCreationStep` at the beginning of the flow, we pass the list of plugins to wait for installation of to the ONBOARD_STORE.

The `waitForPluginInstall` step then uses that list rather than having
special cases in the step for different flows.

If there are no plugins to check, the step is effectively skipped.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/site-creation-step/index.tsx:
```diff
@@ -55,7 +55,11 @@ const SiteCreationStep: Step = function SiteCreationStep( { navigation, flow, da
+	const { setPendingAction, setPluginsToVerify } = useDispatch( ONBOARD_STORE );
+
+	if ( isWooExpressFlow( flow ) ) {
+		setPluginsToVerify( [ 'woocommerce' ] );
```
Should we not run this logic from the flow itself? It might make sense to do it just before we redirect to the `waitForPluginInstall` step.

Having this logic embedded in the step doesn't feel right to me at all!

markbiek replied:
I was going from your earlier comment:

> We could persist the set of plugins at the start of a flow _(or during any step before this one)_, and simply have this step verify that a list of plugin slugs have been installed on the site.

and so I put it in the first step of the flow. But I can definitely move it to the flow itself if you think that makes more sense.

daledupreez replied:
From my side it's more about the separation of concerns than about sequencing, if that makes sense. Having the _step_ responsible for logic that is flow-specific feels like it could easily be fragile. I know we have lots of other logic doing that already, but I think we should aim to minimise that as much as possible.

markbiek replied:
That all makes total sense. I'll get that updated.

markbiek committed:
Move the plugins to check into the flow

Previously, we set the pluginsToVerify in the siteCreationStep. To separate concerns and keep the steps more independent, we now set pluginsToVerify at the beginning of the flow itself.

markbiek committed:
Fix bug checking if we actually have plugins to verify

Our boolean was inverted.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-plugin-install/index.tsx:
```diff
@@ -0,0 +1,117 @@
+		submit?.();
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ siteId ] );
```
Given that the upfront check in the code also depends on `siteSlug`, it may be worth adding it to the deps. Admittedly, one won't change without the other, so this is not blocking.

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/wait-for-plugin-install/index.tsx:
```diff
@@ -0,0 +1,117 @@
+						type: 'plugin_check_timeout',
+						error: 'plugin check took too long',
+						code: 'plugin_check_timeout',
+					} );
```
It might be useful to log the amount of actual elapsed time, or at least the value of `totalTimeout`.

daledupreez commented on client/landing/stepper/declarative-flow/trial-wooexpress-flow.ts:
```diff
@@ -74,7 +76,8 @@ const wooexpress: Flow = {
 
-		const { setStepProgress } = useDispatch( ONBOARD_STORE );
+		const { setStepProgress, setPluginsToVerify } = useDispatch( ONBOARD_STORE );
+		setPluginsToVerify( [ 'woocommerce' ] );
```
Not something for this PR (and maybe something we never need), but we may eventually want to have this data keyed by the flow name. I am not sure if we clear the data across flows, but that might be a situation where we could have subtle data pollution or bugs.
```suggestion
		setPluginsToVerify( flowName, [ 'woocommerce' ] );
```

daledupreez commented on packages/data-stores/src/onboard/reducer.ts:
```diff
@@ -483,6 +483,16 @@ const isMigrateFromWp: Reducer< boolean, OnboardAction > = ( state = false, acti
 
+const pluginsToVerify: Reducer< string[] | undefined, OnboardAction > = ( state, action ) => {
+	if ( action.type === 'SET_PLUGIN_SLUGS_TO_VERIFY' ) {
+		return action.pluginSlugs;
```
Is it worth enforcing that this has type `string[]`? That way we can have `getPluginsToVerify()` be strictly typed to return `undefined | string[]`. (Definitely not blocking, especially seeing as that doesn't seem to be an existing pattern.)

daledupreez reviewed (approved):
LGTM! The flow is pretty slow for me, especially the API call to add the eCommerce trial (which took ~43 seconds!), but the transfer and wait for the plugin to be installed also add up.

We have #73419 open to debug this, but I wanted to note the delays given that they are now longer, even if it is for good reason.

markbiek committed:
Add additional dependcey to useEffect

markbiek commented:
> The flow is pretty slow for me, especially the API call to add the eCommerce trial (which took ~43 seconds!), but the transfer and wait for the plugin to be installed also add up.

It's funny, for me the `assignTrial` step is usually close to instant and the atomic transfer takes forever. 

But agreed that we definitely need to do some work to speed things up!

nelsonec87 reviewed (approved):
LGTM!
I was able to reproduce the problem on `trunk` and, after this patch, created 5 new sites successfully.

markbiek committed:
Log total timeout after plugin check timeout failure
